content/gotraining.slide

```diff
@@ -535,12 +536,45 @@ As part of all labs
 .image _img/crud_puppy.jpg
 
 
-* Errors
+* Errors (1/3)
 
-- [[][Errors]]
-- [[][Panic]]
-- [[][Recover]]
-- [[][Defer]]
+"Errors are values."  - [[https://go-proverbs.github.io/][Rob Pike]]
+
+The `error` type is a built-in interface:
+
+	type error interface {
+	    Error() string
+	}
+
+By convention, functions often return `error` as the second/last return value.
+The calling code should handle errors by testing whether the error equals nil:
+
+	func getCat(strID string) (*Cat, error) {
+		id, err := strconv.Atoi(strID)
+		if err != nil {
+		    return nil, CatError(err)
+		}
+		return catStore.cats[id], nil
+	}
+
+
+* Errors (2/3)
+
+"Don't panic." - [[https://go-proverbs.github.io/][Rob Pike]]
+
+Panic should be used when something went unexpectedly wrong.
+For instance, declare a regular expression as unexported, global `var` and
+initialise it with [[https://golang.org/pkg/regexp/#MustCompile][regexp.MustCompile]] if its pattern is known.
+ `MustCompile` will `panic` at build time if the pattern is invalid.
```

Comment:
this was hard to read. i was confused multiple times. Still thinking of a better way to phrase it.

Reply:
you are right:
how about if I add:
```go
	var re = regexp.MustCompile(`^[a-c]+$`)
	var re2 = regexp.MustCompile(`^[a-c+$`) // panics
```
?

Reply:
How about flipping it:

---

Rule of thumb: Use `panic` rarely and judiciously in your own code!

Golang emphasises a focus on _handling errors explicitly_, if the temptation to use `panic` arises then give strong consideration as to whether your code could be better engineered to avoid it.

Panic should only be used when something critically wrong and neither recovery nor elegant termination are possible.

For example....

---

Reply:
I think any examples are better off showing _scenarios in which you'd panic_ rather than things which panic.

ie. when is it legitimately a good design choice rather than trying to unroll layers of errors?

Reply:
I've tried to reword it. still not great.
i'm too tired for clarity.... ohoh.

Reply:
no, the structure of the paragraph had me confused in multiple places, as well as some odd wording.

I don't understand the relevance of the regular expression being unexported, or global.

It should say "*an* unexported, global var" - with the comma before "global", I read "declare a regular expression as unexported", and the following was another phrase. I spent a few minutes trying to understand that, googling away...

"panic at build time" means, to me, when you build the software.

I started re-writing it but got stuck thinking about just when you should panic, searched the internet for guidance, found none, ate lunch and then thought forget it for now - get your review out - what you have of it...

rethinking now if I can get a moment of Zen.

Reply:
i just learnt that `MustCompile` panics at build time. i thought that was a mistake in your writing. still figuring out how that works...

However, I feel even more now that this is a poor example to use (at least until I understand build-time panics). Most panics will be run time, so I think the "For instance" should be that sort of panic. This `MustCompile` seems like a special case - perhaps still worth calling out - and not be the first example.

Reply:
@pentaphobe - sorry didn't see your comments earlier... i think i also need to wait for a moment Zen.

Reply:
@camh-anz , just wanted to fold in golang regexp best practice here which was probably stupid. `global` so it doesn't have to be re compiled on every function call.
`unexported`, so it can't be modified outside the package.....